File: src/markovchain/wanglandau/energy/WangLandauEnergy.java
package markovchain.wanglandau.energy;

import markovchain.MarkovMove;
import markovchain.wanglandau.WangLandauState;

public abstract class WangLandauEnergy<MarkovState, E extends WangLandauEnergy<MarkovState, E>> {


    public abstract E nextEnergyFromMove(MarkovState state, MarkovMove<MarkovState> moveToNextState);

    public abstract E copy();

    //Must override hashCode() and equals() for HashMap of energies to work correctly
    @Override
    public abstract int hashCode();
    @Override
    public abstract boolean equals(Object o);

    //The user should be able to read the energy
    @Override
    public abstract String toString();

    public abstract static class EnergyFactory<MarkovState, E extends WangLandauEnergy<MarkovState, E>> {

        public abstract E getEnergyFromState(MarkovState state);
    }
}

File: src/markovchain/wanglandau/energy/WangLandauEnergy.kt
package markovchain.wanglandau.energy

import markovchain.MarkovMove
import markovchain.wanglandau.WangLandauMove
import markovchain.wanglandau.WangLandauState

/**
 *
 * @param <MarkovState> The class that represents states in the Markov chain
 * @param <E> The implementation of the specific energy type (added as a generic so the subclassed methods return instances of that specific subclass)
</E></MarkovState> */
abstract class WangLandauEnergy<MarkovState, MM : MarkovMove<MarkovState>, E : WangLandauEnergy<MarkovState, MM, E>> {

    abstract fun copy(): E

    abstract fun getNextEnergyFromMove(move: MM): E


    //Must override hashCode() and equals() for HashMap of energies to work correctly
    abstract override fun hashCode(): Int

    abstract override fun equals(other: Any?): Boolean

    //The user should be able to read the energy
    abstract override fun toString(): String


    abstract class Factory<MarkovState, MM : MarkovMove<MarkovState>, E : WangLandauEnergy<MarkovState, MM, E>> {

        abstract fun getEnergyFromState(state: MarkovState): E
    }
}
